src/psfmachine/machine.py

```diff
@@ -837,289 +635,197 @@ def build_time_model(self, plot=False, downsample=False, split_time_model=False)
-                k &= ~bad_targets
-            #  we save the weights and pixel mask of each segement for later use
-            seg_time_model_w.append(time_model_w)
-            pix_mask_k.append(k)
-
-        self.seg_splits = splits
-        self.time_model_w = np.asarray(seg_time_model_w)
-        self._time_masked = np.asarray(pix_mask_k, dtype=object)
+                for idx in range(self.nt)
+            ]
+        ).transpose((1, 0, 2))
+        flux_norm = flux / np.nanmean(flux, axis=0)
+        flux_err_norm = flux_err / np.nanmean(flux, axis=0)
+
+        # bindown flux arrays
+        flux_binned_raw = P3.bin_func(flux)
+        flux_binned = P3.bin_func(flux_norm)
+        flux_err_binned = P3.bin_func(flux_err_norm)
+
+        # No saturated pixels, 1e5 is a hardcoded value for Kepler.
+        k = (flux_binned_raw < 1e5).all(axis=0)[None, :] * np.ones(
+            flux_binned_raw.shape, bool
+        )
+        # No faint pixels, 100 is a hardcoded value for Kepler.
+        k &= (flux_binned_raw > 100).all(axis=0)[None, :] * np.ones(
+            flux_binned_raw.shape, bool
+        )
+        # No huge variability
+        k &= (np.abs(flux_binned - 1) < 1).all(axis=0)[None, :] * np.ones(
+            flux_binned.shape, bool
+        )
+        # No nans
+        k &= np.isfinite(flux_binned) & np.isfinite(flux_err_binned)
+        k = np.all(k, axis=0)
+        # combine good-behaved pixels with uncontaminated pixels
+        k &= uncontaminated_pixels
+
+        # iterate to remvoe outliers
+        for count in [0, 1, 2]:
+            # need to add pixel_mask=k
+            P3.fit(flux_norm, flux_err=flux_err_norm, pixel_mask=k)
+            res = flux_binned - P3.matrix.dot(P3.weights).reshape(flux_binned.shape)
+            # res = np.ma.masked_array(res, (~k).reshape(flux_binned.shape))
+            res = np.ma.masked_array(res, ~np.tile(k, flux_binned.shape[0]).T)
+            bad_targets = sigma_clip(res, sigma=5).mask
+            bad_targets = bad_targets.any(axis=0)
+            k &= ~bad_targets
+
+        # bookkeeping
+        self.flux_binned = flux_binned
+        self._time_masked_pix = k
+        self.P3 = P3
         if plot:
             return self.plot_time_model()
         return
 
-    def plot_time_model(self, segment=0):
+    def _get_perturbed_model(self, time_indices=None):
         """
-        Diagnostic plot of time model.
+        Computes the perturbed model at given times
+        """
+        if time_indices is None:
+            time_indices = np.arange(len(self.time))
+        time_indices = np.atleast_1d(time_indices)
+        if isinstance(time_indices[0], bool):
+            time_indices = np.where(time_indices[0])[0]
+
+        perturbed_model = []
+        for tdx in time_indices:
+            X = self.mean_model.copy()
+            X.data *= self.P3.model(time_indices=tdx).ravel()
+            perturbed_model.append(X)
+        self.perturbed_model = perturbed_model
```

Comment:
```suggestion
        self.perturbed_model = np.asarray(perturbed_model)
```

Reply:
I don't think this is quite the API we want, do we want something like this:

```python
mac.mean_model
mac.perturbed_model(time_index)
```

because then we can do

```python
mac.perturbed_model(time_index).dot(mac.psf_weights)
```

Right now you're setting an attribute (`self.perturbed_model`) and that's going to waste a lot of memory, and we can't do the dot product above. I think we just want a class method that returns a matrix. Using your code above:

```python

   def perturbed_model(self, time_index=0):
        """
        Computes the perturbed model at given times
        """        
        X = self.mean_model.copy()
        X.data *= self.P3.model(time_indices=time_index).ravel()
        return X
```

What exactly is that ravel doing?

Reply:
That makes sense. 
The `.ravel()` is because `P3.model()` returns a column vector `(1, n_pix)`.